### DISCUSSION TITLE
How to make alembic autogenerate names for foreign_keys?

### DISCUSSION THREAD

**Discussion by maham360** [sqlalchemy/alembic#1024] [Usage Questions]
In my migration file I have alembic bringing up alterations such as:    

```
def upgrade():
        # ### commands auto generated by Alembic - please adjust! ###
        op.create_foreign_key(None, 'complaints', 'notification_history', ['notification_id'], ['id'])
        op.create_foreign_key(None, 'templates_history', 'provider_details', ['provider_id'], ['id'])
        op.create_foreign_key(None, 'templates_history', 'communication_items', ['communication_item_id'], ['id'])
```

I am trying to incorporate this code into my env.py file in order to make alembic name foreign keys automatically, but I'm not sure how to incorporate this into my code.

```py
convention = {
  "ix": "ix_%(column_0_label)s",
  "uq": "uq_%(table_name)s_%(column_0_name)s",
  "ck": "ck_%(table_name)s_%(constraint_name)s",
  "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
  "pk": "pk_%(table_name)s"
}

metadata = MetaData(naming_convention=convention)
```

So far I have it existing in my env.py file like this, but after I run a new migration nothing changes:

```py
from __future__ import with_statement
from alembic import context
from sqlalchemy import engine_from_config, pool, MetaData
from logging.config import fileConfig
from flask import current_app

# this is the Alembic Config object, which provides
# access to the values within the .ini file in use.
config = context.config

# Interpret the config file for Python logging.
# This line sets up loggers basically.
fileConfig(config.config_file_name)

# add your model's MetaData object here
# for 'autogenerate' support


convention = {
  "ix": "ix_%(column_0_label)s",
  "uq": "uq_%(table_name)s_%(column_0_name)s",
  "ck": "ck_%(table_name)s_%(constraint_name)s",
  "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
  "pk": "pk_%(table_name)s"
}

metadata = MetaData(naming_convention=convention)


config.set_main_option(
    'sqlalchemy.url', current_app.config.get('SQLALCHEMY_DATABASE_URI'))
target_metadata = current_app.extensions['migrate'].db.metadata

# other values from the config, defined by the needs of env.py,
# can be acquired:
# my_important_option = config.get_main_option("my_important_option")
# ... etc.


def run_migrations_offline():
    """Run migrations in 'offline' mode.

    This configures the context with just a URL
    and not an Engine, though an Engine is acceptable
    here as well.  By skipping the Engine creation
    we don't even need a DBAPI to be available.

    Calls to context.execute() here emit the given string to the
    script output.

    """
    def include_object(object, name, type_, reflected, compare_to):
        if type_ == "index" or type_== "table" and reflected and compare_to is None:
            return False
        else:
            return True

    url = config.get_main_option("sqlalchemy.url")
    context.configure(
        url=url,
        include_object=include_object,
        target_metadata=target_metadata,)

    with context.begin_transaction():
        context.run_migrations()


def run_migrations_online():
    """Run migrations in 'online' mode.

    In this scenario we need to create an Engine
    and associate a connection with the context.

    """
    def include_object(object, name, type_, reflected, compare_to):
        if type_ == "index" or type_== "table" and reflected and compare_to is None:
            return False
        else:
            return True

    engine = engine_from_config(
                config.get_section(config.config_ini_section),
                prefix='sqlalchemy.',
                poolclass=pool.NullPool)

    connection = engine.connect()
    context.configure(
                connection=connection,
                target_metadata=target_metadata,
                include_object=include_object,
                compare_type=True
                )

    try:
        with context.begin_transaction():
            context.run_migrations()
    finally:
        connection.close()


if context.is_offline_mode():
    run_migrations_offline()
else:
    run_migrations_online()
```

I'm using flask sqlalchemy. Can someone help me get this to work? Or have any better solutions? Thanks

**Comment by zzzeek** [comment]
the naming convention needs to be inside the MetaData object that your model is built on, this one:

> target_metadata = current_app.extensions['migrate'].db.metadata
> 

so wherever Flask sets up the "db.metadata", that's where you set up that naming convention.

here's a stackoverflow example  https://stackoverflow.com/a/29153957/34549

**Reply by maham360** [reply]
Thanks for the reference. Actually, I looked at my model and I see that the `"notification_id"` column has a foreign key:

```
class Complaint(db.Model):
    __tablename__ = 'complaints'

    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    notification_id = db.Column(UUID(as_uuid=True), db.ForeignKey('notification_history.id'),
                                index=True, nullable=False)
```

And id is the primary key so I'm not sure why Alembic is saying there isn't a foreign key for "id". Maybe Alembic isn't detecting the foreign key?

**Reply by zzzeek** [reply]
> ```
> 
> And id is the primary key so I'm not sure why Alembic is saying there isn't a foreign key for "id". Maybe Alembic isn't detecting the foreign key?

this is new information you are introducing here.   how exactly is alembic "saying" this ?

**Reply by maham360** [reply]
Alembic is detecting there is no foreign key for `notification_history`I thought it also detected `id` as needing a foreign key.

**Reply by zzzeek** [reply]
based on the autogenerate illustrated, it indicates there are three foreign key constraints in the model that aren't in the database.  that's what I can see from this end.

**Reply by maham360** [reply]
Ok, that makes sense. I guess I am confused on what this is trying to say exactly:

`op.create_foreign_key(None, 'complaints', 'notification_history', ['notification_id'], ['id'])`

There is no foreign key for the '`complaints`' (source table) or '`notification_history`' (destination table) on the '`notification_id`' column?

**Reply by zzzeek** [reply]
that rule is trying to create a foreign key constraint on the "complaints.notification_id" column of "complaints", referencing "notification_history.id".    

I actually had to look up the API since I dont use this API very much at all, so you can see that at https://alembic.sqlalchemy.org/en/latest/ops.html?highlight=create_foreign_key#alembic.operations.Operations.create_foreign_key

**Reply by maham360** [reply]
Actually I think the issue is something else. I can pass "None" as a foreign key constraint name, it doesn't have to be named.

**Reply by maham360** [reply]
Actually I think that rule is trying to create a foreign key constraint that doesn't exist in the model. Because it's trying to create a foreign key constraint on `complaints.notification_id` and some other column I'm not sure of. I know it's referencing notification_history table

**Reply by zzzeek** [reply]
if the constraints are rendered in the "upgrade" function, then it gets them from the model alone.   they can't be from anywhere else.

**Reply by maham360** [reply]
ok it's creating a foreign key constraint on `complaints.notification_id` and `notification_history.id` (which is a primary key). I'm not sure why this is trying to be done. This constraint doesn't exist

**Reply by maham360** [reply]
I guess my question now is how do I drop foreign key constraints that are in the model (but are not in the database) so alembic doesn't generate these rules: `op.create_foreign_key(None, 'complaints', 'notification_history', ['notification_id'], ['id']) `

**Reply by zzzeek** [reply]
well the first way is to remove the directives from the model.    otherwise, you'd need to configure an exclusion rule using [include_object](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=include_object#alembic.runtime.environment.EnvironmentContext.configure.params.include_object), which is a function that Alembic will ask for every object it's about to generate a revision for.   

but if you expect that the model doesn't have these directives we'd need to look and see why they are there, which if you were to share the class and/or Table definitions for these two tables, we could see why that's happening.

**Reply by maham360** [reply]
Here is the class for the complaints table:

```py
class` Complaint(db.Model):
    __tablename__ = 'complaints'


    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    notification_id = db.Column(UUID(as_uuid=True), db.ForeignKey('notification_history.id'),
                                index=True, nullable=False)
    service_id = db.Column(UUID(as_uuid=True), db.ForeignKey('services.id'), unique=False, index=True, nullable=False)
    service = db.relationship(Service, backref=db.backref('complaints'))
    feedback_id = db.Column(db.Text, nullable=True)
    complaint_type = db.Column(db.Text, nullable=True, default=UNKNOWN_COMPLAINT_TYPE)
    complaint_date = db.Column(db.DateTime, nullable=True)
    created_at = db.Column(db.DateTime, nullable=False, default=datetime.datetime.utcnow)

    def serialize(self):
        return {
            'id': str(self.id),
            'notification_id': str(self.notification_id),
            'service_id': str(self.service_id),
            'service_name': self.service.name,
            'feedback_id': str(self.feedback_id),
            'complaint_type': self.complaint_type,
            'complaint_date': self.complaint_date.strftime(DATETIME_FORMAT) if self.complaint_date else None,
            'created_at': self.created_at.strftime(DATETIME_FORMAT),
        }

```

And here is the class for the notification_history table:

```py
class NotificationHistory(db.Model, HistoryModel):
    __tablename__ = 'notification_history'

    id = db.Column(UUID(as_uuid=True), primary_key=True)
    job_id = db.Column(UUID(as_uuid=True), db.ForeignKey('jobs.id'), index=True, unique=False)
    job = db.relationship('Job')
    job_row_number = db.Column(db.Integer, nullable=True)
    service_id = db.Column(UUID(as_uuid=True), db.ForeignKey('services.id'), index=True, unique=False)
    service = db.relationship('Service')
    template_id = db.Column(UUID(as_uuid=True), index=True, unique=False)
    template_version = db.Column(db.Integer, nullable=False)
    api_key_id = db.Column(UUID(as_uuid=True), db.ForeignKey('api_keys.id'), index=True, unique=False)
    api_key = db.relationship('ApiKey')
    key_type = db.Column(db.String, db.ForeignKey('key_types.name'), index=True, unique=False, nullable=False)
    billable_units = db.Column(db.Integer, nullable=False, default=0)
    notification_type = db.Column(notification_types, index=True, nullable=False)
    created_at = db.Column(db.DateTime, index=True, unique=False, nullable=False)
    sent_at = db.Column(db.DateTime, index=False, unique=False, nullable=True)
    sent_by = db.Column(db.String, nullable=True)
    updated_at = db.Column(db.DateTime, index=False, unique=False, nullable=True, onupdate=datetime.datetime.utcnow)
    status = db.Column(
        'notification_status',
        db.Text,
        db.ForeignKey('notification_status_types.name'),
        index=True,
        nullable=True,
        default='created',
        key='status'  # http://docs.sqlalchemy.org/en/latest/core/metadata.html#sqlalchemy.schema.Column
    )
    reference = db.Column(db.String, nullable=True, index=True)
    client_reference = db.Column(db.String, nullable=True)

    international = db.Column(db.Boolean, nullable=True, default=False)
    phone_prefix = db.Column(db.String, nullable=True)
    rate_multiplier = db.Column(db.Numeric(asdecimal=False), nullable=True)

    created_by_id = db.Column(UUID(as_uuid=True), nullable=True)

    postage = db.Column(db.String, nullable=True)
    status_reason = db.Column(db.String, nullable=True)
    billing_code = db.Column(db.String(256), nullable=True)
    CheckConstraint("""
        CASE WHEN notification_type = 'letter' THEN
            postage is not null and postage in ('first', 'second')
        ELSE
            postage is null
        END
    """)

    __table_args__ = (
        db.ForeignKeyConstraint(
            ['template_id', 'template_version'],
            ['templates_history.id', 'templates_history.version']
        ),
        {}
    )

    @classmethod
    def from_original(cls, notification):
        history = super().from_original(notification)
        history.status = notification.status
        return history

    def update_from_original(self, original):
        super().update_from_original(original)
        self.status = original.status

```

Based on this rule: `op.create_foreign_key(None, 'complaints', 'notification_history', ['notification_id'], ['id'])` I am seeing that `complaints.notification_id` is a foreign key column and `notification_history.id` is the other column that's a primary key, it is trying to create a foreign key constraint on.

**Reply by CaselIT** [reply]
Well the directive is in the model:
```py
notification_id = db.Column(UUID(as_uuid=True), db.ForeignKey('notification_history.id'),
                                index=True, nullable=False)
```

remove the ` db.ForeignKey('notification_history.id')` if you do not want it and alembic will not render that in the update

**Reply by maham360** [reply]
can i manually as a foreign key constraint name in `db.ForeignKey('notification_history.id'` ?

**Reply by CaselIT** [reply]
> can i manually as a foreign key constraint name in `db.ForeignKey('notification_history.id'` ?

sorry, I did not understand the question

**Reply by zzzeek** [reply]
yes you can put a name in there, "db.ForeignKey(..., name="my_name")"

**Reply by maham360** [reply]
Sorry, I meant I wanted to manually add a foreign key constraint name for the relationship between `complaints.notification_id` and `notification_history.id`

**Reply by maham360** [reply]
Ok thank you. I added a fk constraint name and ran another migration instance and this is in the alembic commands:

```
def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_foreign_key('fk_constraint_notification_history', 'complaints', 'notification_history', ['notification_id'], ['id'])
```

It is still showing up as a migration alteration whether I have a name or not, so I guess what I don't understand is why is this happening. And I just want to rid of these, I guess "extra changes", that aren't necessary in my migration file. 

I removed `db.ForeignKey('notification_history.id')` and ran another migration instance and then that upgrade rule did not show up anymore. I'm not sure if I should just remove that or configure a `"include_object"` function to remove those unnecessary mentions somehow.

**Reply by CaselIT** [reply]
> It is still showing up as a migration alteration whether I have a name or not, so I guess what I don't understand is why is this happening. And I just want to rid of these, I guess "extra changes", that aren't necessary in my migration file.

> I removed db.ForeignKey('notification_history.id') and ran another migration instance and then that upgrade rule did not show up anymore. I'm not sure if I should just remove that or configure a "include_object" function to remove those unnecessary mentions somehow.

I really don't understand why you are calling them unnecessary.

The situation is as follows:
- the python model has a foreign key : `db.ForeignKey('notification_history.id')`
- the database has no foreign key
- alembic detects that the there is a fk in the model but not on the database
- a directive to create a foreign key is added by alembic in the migration

Now you have to decide. 
- Is that foreign key needed? Great, then create a migration that adds it.
- It's not needed? Then remove if from the model.

**Reply by maham360** [reply]
I added a ForeignKeyConstraint on the `complaints` table and it's still showing up as "none":

```
    __table_args__ = (
        db.ForeignKeyConstraint(['notification_id'], ['notification_history.id']),
        {}
    )
```

**Reply by CaselIT** [reply]
you can specify a name using `name='something'` in the `ForeignKeyConstraint`

**Reply by maham360** [reply]
Yes I gave it a name but it still shows up in my alembic migration file. I don't think the issue is with the name I think it's the fact that alembic is instructing to create a foreign key in the model that's not in the database and I don't know how to sync them. I'm pretty sure I need the foreign key, I don't want to delete it

**Reply by CaselIT** [reply]
Have you tried keeping the migration that's generated by alembic, and to apply it with `alembic upgrade head` so that it's created in the db?
At that point the db should have the constraint, since it was created by alembic, and autogenerate should no longer detect any change

**Reply by maham360** [reply]
i use flask db migrate to run a new migration and flask db upgrade to apply changes to the database. im not sure why the changes listed here aren't being applied 
```

    op.create_foreign_key('fk_constraint_notification_history', 'complaints', 'notification_history', ['notification_id'], ['id'])
    op.create_unique_constraint('uix_service_callback_channel', 'service_callback', ['service_id', 'callback_channel'])
    op.alter_column('services_history', 'prefix_sms',
               existing_type=sa.BOOLEAN(),
               nullable=False)
    op.create_foreign_key(None, 'templates_history', 'provider_details', ['provider_id'], ['id'])
    op.create_foreign_key(None, 'templates_history', 'communication_items', ['communication_item_id'], ['id'])
    op.drop_constraint('users_identity_provider_user_id', 'users', type_='unique')
```

When I run `flask db upgrade` I get the error:
`[SQL: ALTER TABLE services_history ALTER COLUMN prefix_sms SET NOT NULL]`

Does this mean I need to set the 'prefix_sms' column in the services_history table in my database to NOT NULL? Because services_history table isn't in my model

**Reply by CaselIT** [reply]
What is the database you are using? And what's the complete text of the error, not just the sql being run?

Also I'm not sure how this is related to the above discussion, since that seems a different issue or problem, that's not related with foreign keys

**Reply by maham360** [reply]
Yes it is a different issue. I am using postgres. 

```
ci-db-1          | 2022-05-10 00:31:48.778 UTC [84] ERROR:  column "prefix_sms" contains null values
ci-db-1          | 2022-05-10 00:31:48.778 UTC [84] STATEMENT:  ALTER TABLE services_history ALTER COLUMN prefix_sms SET NOT NULL
ci-migrations-1  | Traceback (most recent call last):
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
ci-migrations-1  |     self.dialect.do_execute(
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute
ci-migrations-1  |     cursor.execute(statement, parameters)
ci-migrations-1  | psycopg2.errors.NotNullViolation: column "prefix_sms" contains null values
ci-migrations-1  | 
ci-migrations-1  | 
ci-migrations-1  | The above exception was the direct cause of the following exception:
ci-migrations-1  | 
ci-migrations-1  | Traceback (most recent call last):
ci-migrations-1  |   File "/usr/local/bin/flask", line 8, in <module>
ci-migrations-1  |     sys.exit(main())
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/flask/cli.py", line 967, in main
ci-migrations-1  |     cli.main(args=sys.argv[1:], prog_name="python -m flask" if as_module else None)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/flask/cli.py", line 586, in main
ci-migrations-1  |     return super(FlaskGroup, self).main(*args, **kwargs)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 782, in main
ci-migrations-1  |     rv = self.invoke(ctx)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
ci-migrations-1  |     return _process_result(sub_ctx.command.invoke(sub_ctx))
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
ci-migrations-1  |     return _process_result(sub_ctx.command.invoke(sub_ctx))
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
ci-migrations-1  |     return ctx.invoke(self.callback, **ctx.params)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
ci-migrations-1  |     return callback(*args, **kwargs)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
ci-migrations-1  |     return f(get_current_context(), *args, **kwargs)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/flask/cli.py", line 426, in decorator
ci-migrations-1  |     return __ctx.invoke(f, *args, **kwargs)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
ci-migrations-1  |     return callback(*args, **kwargs)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/flask_migrate/cli.py", line 134, in upgrade
ci-migrations-1  |     _upgrade(directory, revision, sql, tag, x_arg)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/flask_migrate/__init__.py", line 95, in wrapped
ci-migrations-1  |     f(*args, **kwargs)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/flask_migrate/__init__.py", line 280, in upgrade
ci-migrations-1  |     command.upgrade(config, revision, sql=sql, tag=tag)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/command.py", line 320, in upgrade
ci-migrations-1  |     script.run_env()
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/script/base.py", line 563, in run_env
ci-migrations-1  |     util.load_python_file(self.dir, "env.py")
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/util/pyfiles.py", line 92, in load_python_file
ci-migrations-1  |     module = load_module_py(module_id, path)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/util/pyfiles.py", line 108, in load_module_py
ci-migrations-1  |     spec.loader.exec_module(module)  # type: ignore
ci-migrations-1  |   File "<frozen importlib._bootstrap_external>", line 843, in exec_module
ci-migrations-1  |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
ci-migrations-1  |   File "migrations/env.py", line 117, in <module>
ci-migrations-1  |     run_migrations_online()
ci-migrations-1  |   File "migrations/env.py", line 110, in run_migrations_online
ci-migrations-1  |     context.run_migrations()
ci-migrations-1  |   File "<string>", line 8, in run_migrations
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/runtime/environment.py", line 851, in run_migrations
ci-migrations-1  |     self.get_context().run_migrations(**kw)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/runtime/migration.py", line 620, in run_migrations
ci-migrations-1  |     step.migration_fn(**kw)
ci-migrations-1  |   File "/app/migrations/versions/d7a44e14e03e_.py", line 21, in upgrade
ci-migrations-1  |     op.alter_column('services_history', 'prefix_sms',
ci-migrations-1  |   File "<string>", line 8, in alter_column
ci-migrations-1  |   File "<string>", line 3, in alter_column
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/operations/ops.py", line 1880, in alter_column
ci-migrations-1  |     return operations.invoke(alt)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/operations/base.py", line 392, in invoke
ci-migrations-1  |     return fn(self, operation)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/operations/toimpl.py", line 50, in alter_column
ci-migrations-1  |     operations.impl.alter_column(
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/ddl/postgresql.py", line 170, in alter_column
ci-migrations-1  |     super(PostgresqlImpl, self).alter_column(
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/ddl/impl.py", line 227, in alter_column
ci-migrations-1  |     self._exec(
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/alembic/ddl/impl.py", line 193, in _exec
ci-migrations-1  |     return conn.execute(construct, multiparams)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1011, in execute
ci-migrations-1  |     return meth(self, multiparams, params)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/sql/ddl.py", line 72, in _execute_on_connection
ci-migrations-1  |     return connection._execute_ddl(self, multiparams, params)
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1068, in _execute_ddl
ci-migrations-1  |     ret = self._execute_context(
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1316, in _execute_context
ci-migrations-1  |     self._handle_dbapi_exception(
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1510, in _handle_dbapi_exception
ci-migrations-1  |     util.raise_(
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
ci-migrations-1  |     raise exception
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1276, in _execute_context
ci-migrations-1  |     self.dialect.do_execute(
ci-migrations-1  |   File "/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 608, in do_execute
ci-migrations-1  |     cursor.execute(statement, parameters)
ci-migrations-1  | sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) column "prefix_sms" contains null values
ci-migrations-1  | 
ci-migrations-1  | [SQL: ALTER TABLE services_history ALTER COLUMN prefix_sms SET NOT NULL]
```

This is the complete error message. I tried setting `server_default=None` in the newly created migrations file under def upgrade:

```
def upgrade():
               op.alter_column('services_history', 'prefix_sms',
               server_default=None,
               existing_type=sa.BOOLEAN(),
               nullable=False)
```

But it's still throwing the error when I try to run a new migration. I'm using docker compose to create a new migration. I don't know if the issue is with docker or what at this point. `ci-migrations-1` and `ci-db-1` are both containers.

**Reply by CaselIT** [reply]
Again the error is telling you the problem: `NotNullViolation: column "prefix_sms" contains null values`

- You have set `prefix_sms` in your model from `nullable=True` to `nullable=False`.
- Alembic detected that and created the migration
- The table in the db has null values in that column, so that cannot be null until you have not updated these values

You have to decide how to proceed here.
If that column needs to be not nullable, an option is to change the null value to a default if it has one, by running an update in the migration before the `op.aleter_column`. But it depends on your application, there is no problem I can see here from the alembic point of view

**Reply by maham360** [reply]
I see. So I want `nullable=False` for this column. I ran an update in an older migration file where `'services_history's'`  `'prefix_sms'` column was first created, and it had `nullable=True`. So in that rule in the older file I added `server_deafault` here:

```
def upgrade():
    op.add_column('services_history', sa.Column('prefix_sms', sa.Boolean(), nullable=True, server_default='t'))
```
After doing so, I was able to get around the `[SQL: ALTER TABLE services_history ALTER COLUMN prefix_sms SET NOT NULL]` error. I mean I guess this technically works, but I don't want the `prefix_sms` column to have a default value of `'true'` so I guess I am stuck on what to have the default value as. I want to forbid nulls if a column fails if some rows contain null already.

**Reply by CaselIT** [reply]
you can create a new migration or update the current one to remove the server default after adding the column.

**Reply by maham360** [reply]
Well I added `server_default='t'` to get past the SQL error, so I can just remove it from the old migration I added it to. That wouldn't really solve my issue, I'd have that SQL ALTER TABLE error again.

**Reply by CaselIT** [reply]
The process I follow to add a non nullable column is
- add the column with a server default `op.add_column`
- remove the server default `op.alter_column`
